test-runtime/build.rs

```diff
@@ -0,0 +1,152 @@
+    // Select substrate binary to run based on env var.
+    let substrate_bin =
+        env::var(SUBSTRATE_BIN_ENV_VAR).unwrap_or_else(|_| "substrate".to_owned());
+
+    // Run binary.
+    let port = next_open_port()
+        .expect("Cannot spawn substrate: no available ports in the given port range");
+    let cmd = Command::new(&substrate_bin)
+        .arg("--dev")
+        .arg("--tmp")
+        .arg(format!("--rpc-port={}", port))
+        .spawn();
+    let mut cmd = match cmd {
+        Ok(cmd) => cmd,
+        Err(e) => {
+            panic!("Cannot spawn substrate command '{}': {}", substrate_bin, e)
+        }
+    };
+
+    // Download metadata from binary; retry until successful, or a limit is hit.
+    let res: serde_json::Value = {
+        const MAX_RETRIES: usize = 20;
+        let mut retries = 0;
+        let mut wait_secs = 1;
+        loop {
+            if retries >= MAX_RETRIES {
+                panic!("Cannot connect to substrate node after {} retries", retries);
+            }
+            let res = reqwest::blocking::Client::new()
+                .post(format!("http://localhost:{}", port))
+                .json(&json!({
+                   "id": 1,
+                   "jsonrpc": "2.0",
+                   "method": "state_getMetadata",
+                }))
+                .send();
+            match res {
+                Ok(res) if res.status().is_success() => {
+                    let _ = cmd.kill();
+                    break res
+                        .json()
+                        .expect("valid JSON response from substrate node expected")
+                }
+                _ => {
+                    thread::sleep(time::Duration::from_secs(wait_secs));
+                    retries += 1;
+                    wait_secs += 1;
+                }
+            };
+        }
+    };
+    let metadata_hex = res["result"]
+        .as_str()
+        .expect("Metadata should be returned as a string of hex encoded SCALE bytes");
+    let metadata_bytes = hex::decode(&metadata_hex.trim_start_matches("0x")).unwrap();
+
+    // Save metadata to a file:
+    let out_dir = env::var_os("OUT_DIR").unwrap();
+    let metadata_path = Path::new(&out_dir).join("metadata.scale");
+    fs::write(&metadata_path, metadata_bytes).expect("Couldn't write metadata output");
+
+    // Write out our expression to generate the runtime API to a file. Ideally, we'd just write this code
+    // in lib.rs, but we must pass a string literal (and not `concat!(..)`) as an arg to runtime_metadata_path,
+    // and so we need to spit it out here and include it verbatim instead.
+    let runtime_api_contents = format!(
+        "
+        #[subxt::subxt(
+            runtime_metadata_path = \"{}\",
+            generated_type_derives = \"Debug, Eq, PartialEq\"
+        )]
+        pub mod node_runtime {{
+            #[subxt(substitute_type = \"sp_arithmetic::per_things::Perbill\")]
+            use sp_runtime::Perbill;
+        }}
+    ",
```

Comment:
Could use a raw string literal here (`r#".."#`) to avoid the escapes.

Reply:
I need to format in the string path though :)

Reply:
I thought it did...one way to find out...

Reply:
Ah, indeed the raw string does work! I'm not sure why I thoguht it didn't now..